sql-plugin/src/main/scala/com/nvidia/spark/rapids/io/async/AsyncResources.scala

```diff
@@ -0,0 +1,442 @@
+  }
+}
+
+/**
+ * Result wrapper for AsyncRunner execution that carries the task's output data and optional
+ * resource release callback for deferred resource management.
+ */
+trait AsyncResult[T] extends AutoCloseable {
+  val data: T
+
+  /**
+   * Metrics associated with the task execution, such as scheduling and execution time.
+   */
+  val metrics: AsyncMetrics
+
+  /**
+   * Indicates whether the result holds a release hook to be executed upon closure.
+   */
+  val releaseHook: Option[() => Unit] = None
+
+  /**
+   * Trigger release hook if it exists. Do not try to close the data itself, because
+   * the lifecycle of the data is managed by the caller.
+   */
+  override def close(): Unit = {
+    releaseHook.foreach(callback => callback())
+  }
+}
+
+class SimpleAsyncResult[T](override val data: T,
+    override val metrics: AsyncMetrics) extends AsyncResult[T]
+
+class AsyncResultDecayRelease[T](override val data: T,
+    override val metrics: AsyncMetrics,
+    override val releaseHook: Option[() => Unit]) extends AsyncResult[T]
+
+/**
+ * The AsyncRunner interface represents a resource-aware task that can be scheduled by
+ * ResourceBoundedThreadExecutor. Tasks define their resource requirements, execution priority,
+ * and can optionally hold resources after completion for deferred release.
+ *
+ * AsyncRunner provides hooks for resource lifecycle management and supports both immediate
+ * and deferred resource release patterns based on the task's holdResourceAfterCompletion flag.
+ */
+trait AsyncRunner[T] extends Callable[AsyncResult[T]] {
+  /**
+   * Resource required by the task, such as host memory or GPU semaphore.
+   */
+  def resource: TaskResource
+
+  /**
+   * Priority of the async runner, higher value means higher priority.
+   *
+   * NOTE: The priority may be the same as the universal priority generated by TaskPriority under
+   * most circumstances, such as GpuMultiFileReader, but it can also be adjusted based on the
+   * resource requirements or other factors.
+   */
+  def priority: Double
+
+  /**
+   * The abstract method defines the actual task logic, which should be implemented by the
+   * subclass.
+   */
+  protected def callImpl(): T
+
+  /**
+   * Builds the result of the task execution, which includes the data and a callback to release
+   * resources if needed. This method can be overridden by subclasses to customize the result
+   * format to carry additional metadata or state.
+   */
+  protected def buildResult(resultData: T, metrics: AsyncMetrics): AsyncResult[T] = {
+    if (!holdResourceAfterCompletion) {
+      new SimpleAsyncResult(resultData, metrics)
+    } else {
+      new AsyncResultDecayRelease(resultData, metrics,
+        releaseHook = Option(() => this.callDecayReleaseCallback()))
+    }
+  }
+
+  def call(): AsyncResult[T] = {
+    val startTime = System.nanoTime()
+    val resultData = try {
+      beforeExecuteHook.foreach { hook => hook() }
+      callImpl()
```

Comment:
in the event of an exception in `callImpl` (say the runner itself failed). Who is going to release the resource?